hls-plugin-api/src/Ide/Plugin/RangeMap.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+-- | A map that allows fast \"in-range\" filtering. 'RangeMap' is meant
+-- to be constructed once and cached as part of a Shake rule. If
+-- not, the map will be rebuilt upon each invocation, yielding slower
+-- results compared to the list-based approach!
+module Ide.Plugin.RangeMap
+  ( RangeMap(..),
+    fromList,
+    fromList',
+    filterByRange,
+  ) where
+
+import           Data.Bifunctor                           (first)
+import           Data.Foldable                            (foldl')
+import           Development.IDE.Graph.Classes            (NFData)
+import qualified HaskellWorks.Data.IntervalMap.FingerTree as IM
+import           Language.LSP.Types                       (Position,
+                                                           Range (Range))
+
+-- | A map from code ranges to values.
+newtype RangeMap a = RangeMap
+  { unRangeMap :: IM.IntervalMap Position a
+    -- ^ 'IM.Interval' of 'Position' corresponds to a 'Range'
+  }
+  deriving newtype (NFData)
```

Comment:
might as well add `Show, Eq, Ord`?

Reply:
I only added what was currently necessary at the use-site, but adding those as well seems like a good idea :+1:

Reply:
Well, `IntervalMap` apparently [doesn't have](https://hackage.haskell.org/package/hw-fingertree-0.1.2.1/docs/HaskellWorks-Data-IntervalMap-FingerTree.html#t:IntervalMap) those instances. Perhaps it would make sense to provide the `Functor` instance though?

Reply:
Huh. I think maybe I just scrolled to `Interval` by accident :sweat_smile: `Traversable` and `Foldable` and `Monoid` are probably all useful too.